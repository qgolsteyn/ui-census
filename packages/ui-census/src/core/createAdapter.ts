import { Dict } from "../types";
import { CensusObject, CensusDefinition } from "./types";
import { createBaseProxyHandler } from "./utils/proxy";
import { querySync } from "./query/querySync";

const createQueryProxy = <ElementType>(
  element: ElementType,
  queryResolver: Dict<(element: ElementType) => any>,
  actionResolver: Dict<(element: ElementType) => any>
) => {
  const combinedResolvers = { ...queryResolver, ...actionResolver };
  const handler: ProxyHandler<Dict> = {
    ...createBaseProxyHandler(Reflect.ownKeys(queryResolver)),
    get: (_, p) => {
      if (typeof p === "string" && p in combinedResolvers) {
        return combinedResolvers[p](element);
      } else {
        return undefined;
      }
    },
  };

  return new Proxy({} as any, handler);
};

const createAdapter = <
  ElementType,
  Definition extends CensusDefinition<ElementType>
>(
  definition: Definition
) => (target: ElementType) => {
  const doc: CensusObject<Definition> = {} as any;

  for (const key in definition) {
    doc[key] = (...args: any[]) => {
      return querySync(
        definition[key]
          .selector(target, args)
          .map((element) =>
            createQueryProxy(
              element,
              definition[key].queries,
              definition[key].actions
            )
          )
      );
    };
  }

  return doc;
};

export default createAdapter;
