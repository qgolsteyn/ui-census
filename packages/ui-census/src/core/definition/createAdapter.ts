import { Dict, Serializeable, Index } from "../../types";
import { ElementAccessor } from "../types";
import { querySync } from "../query/querySync";
import { createBaseProxyHandler } from "../utils/proxy";

const createQueryProxy = <
  ElementType,
  Queries extends { [key: string]: (element: ElementType) => Serializeable },
  Actions extends {
    [key: string]: (element: ElementType) => (...args: any[]) => any;
  }
>(
  element: ElementType,
  queryResolver: Queries,
  actionResolver: Actions
) => {
  const combinedResolvers = { ...queryResolver, ...actionResolver };
  const handler: ProxyHandler<Dict> = {
    get: (obj, p) => {
      if (typeof p === "string" && p in combinedResolvers) {
        return combinedResolvers[p](element);
      } else {
        return obj[p as any];
      }
    },
  };

  return new Proxy(
    {},
    { ...handler, ...createBaseProxyHandler(Object.keys(queryResolver)) }
  ) as {
    [Key in keyof Queries]: ReturnType<Queries[Key]>;
  } &
    { [Key in keyof Actions]: ReturnType<Actions[Key]> };
};

const createAdapter = <
  ElementTypeIn,
  ElementTypeOut,
  QueryParameters extends Array<any>,
  Queries extends { [key: string]: Serializeable },
  Actions extends {
    [key: string]: (...args: any[]) => any;
  }
>(
  selector: (
    target: ElementTypeIn,
    ...args: QueryParameters
  ) => ElementTypeOut[],
  queries: {
    [Key in keyof Queries]: (element: ElementTypeOut) => Queries[Key];
  },
  actions: {
    [Key in keyof Actions]: (element: ElementTypeOut) => Actions[Key];
  }
) => (
  target: ElementTypeIn
): ElementAccessor<QueryParameters, Queries, Actions> => (
  ...args: QueryParameters
) => {
  const queryProxies = selector(target, ...args).map((element) =>
    createQueryProxy(element, queries, actions)
  );

  return querySync(queryProxies, Object.keys(queries), Object.keys(actions));
};

export default createAdapter;
