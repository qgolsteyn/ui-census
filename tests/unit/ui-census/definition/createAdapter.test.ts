import createAdapter from "ui-census/src/core/definition/createCensusObjectFactory";

const wordAdapter = createAdapter(
  (target: string) => {
    let elements = target.split(" ");

    return elements;
  },
  {
    text: (element) => element,
    length: (element) => element.length,
  },
  {
    getElement: (element) => () => element,
  }
);

test("it handles a basic query", () => {
  const wordAccessor = wordAdapter("hello");

  const word = wordAccessor().single();

  expect(word.text).toBe("hello");
  expect(word.length).toBe(5);
});

test("it handles a query by prop with mapping", () => {
  const wordAccessor = wordAdapter("this is a test");

  const divObjects = wordAccessor().all();

  expect(divObjects.map((node) => node.text)).toStrictEqual([
    "this",
    "is",
    "a",
    "test",
  ]);
});

test("it handles object deep equality", () => {
  const wordAccessor = wordAdapter("canadian");

  const divObject = wordAccessor().single();

  expect(Object.keys(divObject)).toStrictEqual(["text", "length"]);
  expect(divObject).toEqual({ text: "canadian", length: 8 });
});

test("it supports snapshot testing", () => {
  const wordAccessor = wordAdapter("snapshot testing");

  expect(wordAccessor().all()).toMatchSnapshot();
  expect(wordAccessor().grouped()).toMatchSnapshot();
});

test("it handles a simple filter", () => {
  const wordAccessor = wordAdapter("Lorem ipsum dolor sit amet");

  const match = wordAccessor().contains({ text: "amet" }).single();

  expect(match).toMatchObject({ length: 4 });
});
