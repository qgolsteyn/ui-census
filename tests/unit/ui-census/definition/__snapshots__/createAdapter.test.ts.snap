// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it supports snapshot testing 1`] = `
Array [
  Object {
    "length": 8,
    "text": "snapshot",
  },
  Object {
    "length": 7,
    "text": "testing",
  },
]
`;

exports[`it supports snapshot testing 2`] = `
Object {
  "length": Array [
    8,
    7,
  ],
  "text": Array [
    "snapshot",
    "testing",
  ],
}
`;
